usersession/agent/rest_api.go

```diff
@@ -460,11 +463,43 @@ func postRefreshFinishedNotification(c *Command, r *http.Request) Response {
 		notification.WithDesktopEntry("io.snapcraft.SessionAgent"),
 		notification.WithUrgency(notification.LowUrgency),
 	}
+	si, err := snap.ReadCurrentInfo(finishRefresh.InstanceName)
+
+	parser := goconfigparser.New()
+	if err == nil {
+		// trivial heuristic, if the app is named like a snap then
+		// it's considered to be the main user facing app and hopefully carries
+		// a nice icon
+		mainApp, ok := si.Apps[si.SnapName()]
+		if ok && !mainApp.IsService() {
+			// got the main app, grab its desktop file
+			if err := parser.ReadFile(mainApp.DesktopFile()); err == nil {
+				icon, _ = parser.Get("Desktop Entry", "Icon")
+			}
+		}
+	} else {
+		logger.Noticef("cannot load snap-info for %s: %v", finishRefresh.InstanceName, err)
+	}
+
+	if icon == "" {
+		// If it doesn't exist, take the first app in the snap with a DesktopFile with icon
+		for _, app := range si.Apps {
+			if app.IsService() {
+				continue
+			}
+			if err := parser.ReadFile(app.DesktopFile()); err == nil {
+				if icon, err = parser.Get("Desktop Entry", "Icon"); (icon != "") && (err == nil) {
+					break
+				}
+			}
+		}
+	}
```

Comment:
I think it grew a bit, how about moving the piece that finds icons to a separate function?

```
func guessAppIcon(si *snap.Info) (icon string) {
	parser := goconfigparser.New()
	// trivial heuristic, if the app is named like a snap then
	// it's considered to be the main user facing app and hopefully carries
	// a nice icon
	mainApp, ok := si.Apps[si.SnapName()]
	if ok && !mainApp.IsService() {
		// got the main app, grab its desktop file
		if err := parser.ReadFile(mainApp.DesktopFile()); err == nil {
			icon, _ = parser.Get("Desktop Entry", "Icon")
		}
	}
	if icon != "" {
		return icon
	}

	// If it doesn't exist, take the first app in the snap with a DesktopFile with icon
	for _, app := range si.Apps {
		if app.IsService() {
			continue
		}
		if err := parser.ReadFile(app.DesktopFile()); err != nil {
			continue
		}
		var err error
		if icon, err = parser.Get("Desktop Entry", "Icon"); err == nil && icon != "" {
			return icon
		}
	}
	return ""
}
```
then the code here becomes easy to follow:

```suggestion
	si, err := snap.ReadCurrentInfo(finishRefresh.InstanceName) {
		icon = guessAppIcon(si)
	} else {
		logger.Noticef("cannot load snap-info for %s: %v", finishRefresh.InstanceName, err)
	}
```

Please add unit tests in rest_api_test.go as well. You'll likely need to mock the snaps and their desktop files, so you can start with something like:

```
func (s *restSuite) TestPostCloseRefreshNotificationWithIconDefault(c *C) {
	// add a notification first
	dirs.SetRootDir(c.MkDir())
	
	mockYaml := `
name: some-app
apps:
  other-app:
    command: /bin/foo
  some-app:
    command: /bin/foo
`
	snaptest.MockSnapCurrent(c, mockYaml[1:], &snap.SideInfo{
		Revision: snap.R("42"),
	})

	refreshInfo := &client.PendingSnapRefreshInfo{InstanceName: "some-snap"}
	s.testPostPendingRefreshNotificationBody(c, refreshInfo)

	desktopEntry := `
[Desktop Entry]
Icon=foo.png
`

	c.Assert(os.WriteFile(filepath.Join(dirs.SnapDesktopFilesDir, "some-snap_some-snap.desktop"), desktopEntry[1:]), 
		NotNil)
	notifications := s.notify.GetAll()
	c.Assert(notifications, HasLen, 1)
	n := notifications[0]
	// boring stuff is checked above
	c.Check(n.Summary, Equals, `some-snap was updated.`)
	c.Check(n.Body, Equals, "Ready to launch.")
	c.Check(n.Hints, DeepEquals, map[string]dbus.Variant{
		"urgency":       dbus.MakeVariant(byte(notification.LowUrgency)),
		"desktop-entry": dbus.MakeVariant("io.snapcraft.SessionAgent"),
		"icon": dbus.MakeVariant("foo.png"),
	})
}
```

Reply:
Done. I did the tests only for the new function, guessAppIcon().

Reply:
It would be a good idea to have some walkthrough tests like @bboozzoo mentions above to ensure this works e2e.

Reply:
I don't see it. I mean... those tests wouldn't be for searching the icon, but for the notifications themselves, so I think that they should be in a different MR...

Reply:
Done.